mp4/lou.go

```diff
@@ -0,0 +1,279 @@
+package mp4
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/Eyevinn/mp4ff/bits"
+)
+
+// TlouBox - Track loudness info Box
+//
+// Contained in : Ludt Box (ludt)
+type TlouBox struct {
+	loudnessBaseBox
+}
+
+// TlouBox - Album loudness info Box
+//
+// Contained in : Ludt Box (ludt)
+type AlouBox struct {
+	loudnessBaseBox
+}
+
+type LoudnessBase struct {
+	EQSetID                uint8
+	DownmixID              uint8
+	DRCSetID               uint8
+	BsSamplePeakLevel      int16
+	BsTruePeakLevel        int16
+	MeasurementSystemForTP uint8
+	ReliabilityForTP       uint8
+	Measurements           []*Measurement
+}
+
+type Measurement struct {
+	MethodDefinition  uint8
+	MethodValue       uint8
+	MeasurementSystem uint8
+	Reliability       uint8
+}
+
+// loudnessBaseBox according to ISO/IEC 14496-12 Section 12.2.7.2
+type loudnessBaseBox struct {
+	Version       byte
+	Flags         uint32
+	LoudnessBases []*LoudnessBase
+}
+
+func decodeLoudnessBaseBoxSR(hdr BoxHeader, startPos uint64, sr bits.SliceReader) (*loudnessBaseBox, error) {
+	versionAndFlags := sr.ReadUint32()
+	version := byte(versionAndFlags >> 24)
+	b := &loudnessBaseBox{
+		Version: version,
+		Flags:   versionAndFlags & flagsMask,
+	}
+	var loudnessBaseCount uint8
+	if b.Version >= 1 {
+		loudnessBaseCount = 0x3f & sr.ReadUint8()
+	} else {
+		loudnessBaseCount = 1
+	}
+
```

Comment:
Since the length of the loudnessBases slice is known, it is more efficient to preallocate it with the right size. In this particular case, the size is typically 1, so the gain is small, but it is a good general principle.